src/custom-elements.js

```diff
@@ -635,6 +646,21 @@ let Deferred;
     }
   }
 
+  /**
+   * @param {!Map<string, !CustomElementDefinition>} definitions
+   * @param {!Node|!HTMLElement|null} node
+   */
+  function getDefinition(definitions, node) {
+    const name = typeof node.is === 'string' ? node.is : node.tagName.toLowerCase();
+    const definition = definitions.get(name);
+    if (definition) {
+      // Make sure local name matches the actual node's tagName
+      return definition.localName === node.tagName.toLowerCase() || definition.localName === node.is ? definition : null;
```

Comment:
We want to return null in the following case:

```js
customElements.define('x-foo', class XFoo extends HTMLButtonElement {}, {extends: 'button'});
document.createElement('div', {is: 'x-foo'}); // This is not a button so the element is not upgraded to an x-foo.
```

See https://github.com/webcomponents/custom-elements/pull/42/files#diff-ba6f9bc9f409ee1e78139fbd7d15902eR283